icepyx/tests/test_granules.py

```diff
@@ -592,14 +594,40 @@ def test_no_granules_in_search_results():
 
 def test_correct_granule_list_returned():
     reg_a = ipx.Query(
-        "ATL06", [-55, 68, -48, 71], ["2019-02-20", "2019-02-28"], version="3",
+        "ATL06",
+        [-55, 68, -48, 71],
+        ["2019-02-20", "2019-02-28"],
+        version="3",
     )
 
-    obs_grans, = reg_a.avail_granules(ids=True)
+    (obs_grans,) = reg_a.avail_granules(ids=True)
     exp_grans = [
         "ATL06_20190221121851_08410203_003_01.h5",
         "ATL06_20190222010344_08490205_003_01.h5",
         "ATL06_20190225121032_09020203_003_01.h5",
         "ATL06_20190226005526_09100205_003_01.h5",
     ]
     assert set(obs_grans) == set(exp_grans)
+
+
+@responses.activate
+def test_avail_granule_CMR_error():
+    badreq = "https://cmr.earthdata.nasa.gov/search/granules.json?version=003&temporal=badinput&short_name=ATL08"
+    responses.add(
+        responses.GET,
+        badreq,
+        json={
+            "errors": "temporal start datetime is invalid: [badinput] is not a valid datetime."
+        },
+    )
+
+    response = requests.get(badreq)
+
+    assert response.json() == {
+        "errors": "temporal start datetime is invalid: [badinput] is not a valid datetime."
+    }
+    assert responses.calls[0].request.url == badreq
+    assert (
+        responses.calls[0].response.text
+        == '{"errors": "temporal start datetime is invalid: [badinput] is not a valid datetime."}'
+    )
```

Comment:
You have the right idea, but just need to change the assertions to test the icepyx code.

These assertions are demonstrating what the `responses` library does, which is to make `requests.get` react however you specify just for this test, rather than actually making the request on the web. The url in `responses.add`, e.g. `badreq` is what `responses` is looking for everytime `requests.get` is called, and if it matches it will return the response you configured.

What we want is if you call `Granules.get_avail` in this test, it will use the 'mocked' request. To do that, the url here (`badreq`) needs to match what would be called by `Granules.get_avail` for the mock to recognize and return the bad response. In this case it could probably just be a catchall regular expression like `re.compile(re.escape('http://cmr.earthdata.nasa.gov/search/granules.json') + r'.*')` so that any cmr granule search request in duration of this test returns the error response you specified. Does that make sense? E.g.:
```
responses.add(
    responses.GET,
    re.compile(re.escape('http://cmr.earthdata.nasa.gov/search/granules.json') + r'.*'),
    status=400,
    json={
        "errors": "temporal start datetime is invalid: [badinput] is not a valid datetime."
    },
)
```

Note the `status=400` so the mocked response has that status code. After that, if setup a Granules object and call get_avail(), it should trigger the error and you can, for example, `assert_raises(NsidcQueryError)` to see that your code does the thing you expect.

Reply:
I've gotten like 95% of the way there, but I still can't get the test to pass... there's clearly some string matching formatting something I'm missing, but the failed assertion shows nothing obvious to my eye. Any idea?

Reply:
Ah, in `pytest.raises` the `match` argument is interpreted as a regex and as such the square brackets have a special meaning.  To "escape" (signal that they should be treated as literal square brackets) you can do:
```with pytest.raises(AssertionError, match=re.escape(ermsg)):```

Reply:
Ah, in `pytest.raises` the `match` argument is interpreted as a regex and as such the square brackets have a special meaning.  To "escape" (signal that they should be treated as literal square brackets) you can do:
```with pytest.raises(AssertionError, match=re.escape(ermsg)):```

Reply:
Thank you - worked perfectly! Nic and I played a similar game with Window's paths in ICESat-2 data yesterday.